sympy/polys/rootoftools.py

```diff
@@ -772,15 +775,64 @@ def _get_roots(cls, method, poly, radicals):
         free_names = {str(i) for i in poly.free_symbols}
         for x in chain((symbols('x'),), numbered_symbols('x')):
             if x.name not in free_names:
-                poly = poly.xreplace({d: x})
+                poly = poly.replace(d, x)
                 break
+
+        if dom.is_QQ or dom.is_ZZ:
+            return cls._get_roots_qq(method, poly, radicals)
+        elif dom.is_AlgebraicField or dom.is_ZZ_I or dom.is_QQ_I:
+            return cls._get_roots_alg(method, poly, radicals)
+        else:
+            # XXX: not sure how to handle ZZ[x] which appears in some tests?
+            # this makes the tests pass alright but has to be a better way?
+            return cls._get_roots_qq(method, poly, radicals)
+
+
+    @classmethod
+    def _get_roots_qq(cls, method, poly, radicals):
+        """Return postprocessed roots of specified kind
+         for polynomials with rational coefficients. """
         coeff, poly = cls._preprocess_roots(poly)
         roots = []
 
         for root in getattr(cls, method)(poly):
             roots.append(coeff*cls._postprocess_root(root, radicals))
+
         return roots
 
+    @classmethod
+    def _get_roots_alg(cls, method, poly, radicals):
+        """Return postprocessed roots of specified kind
+         for polynomials with algebraic coefficients. It assumes
+         the domain is already an algebraic field. First it
+         finds the roots using _get_roots_qq, then uses the
+         square-free factors to filter roots and get the correct
+         multiplicity.
+         """
+
+        # Existing QQ code can find and sort the roots
+        roots = cls._get_roots_qq(method, poly.lift(), radicals)
+
+        subroots = {}
+        for f, m in poly.sqf_list()[1]:
+            if method == "_real_roots":
+                roots_filt = f.which_roots(roots, real=True)
+            elif method == "_all_roots":
+                roots_filt = f.which_roots(roots, real=False)
+            for r in roots_filt:
+                if r in subroots:
+                    subroots[r] += m
+                else:
+                    subroots[r] = m
+
+        # loop over unique roots, keep ones in subroots, and store multiplicity
+        roots_filtered_mult = [(r, subroots[r]) for r in list(dict.fromkeys(roots)) if r in subroots]
+
+        # flatten the roots list according to multiplicity
+        roots_flat = sum(([r] * m for r, m in roots_filtered_mult), [])
```

Comment:
We can make this simpler:
```python
roots_flat = []
for r in roots:
    if r in subroots:
        m = subroots[r]
        roots_flat.extend([r] * m)
```

Reply:
Sounds good, although we have to add a way to not add the same root again to roots_flat. Eg with this code we get
```
In [5]: Poly((x-1) * (x-sqrt(2))**2, x, extension=True).real_roots()
Out[5]: [1, 1, sqrt(2), sqrt(2), sqrt(2), sqrt(2)]
```
Bc the lift has roots `[-sqrt(2), -sqrt(2), 1, 1, sqrt(2), sqrt(2)]`, so the sqrt(2) gets looped over twice (and mult is 2, so we get 4 of them in the final output). 

I've added a check for this to your code.

```python
        roots_seen = set()
        roots_flat = []
        for r in roots:
            if r in subroots and r not in roots_seen:
                m = subroots[r]
                roots_flat.extend([r] * m)
                roots_seen.add(r)
```